src/libraries/System.Net.Quic/tests/FunctionalTests/QuicTestBase.cs

```diff
@@ -52,24 +54,92 @@ internal QuicListener CreateQuicListener(IPEndPoint endpoint)
             return listener;
         }
 
-        internal async Task RunClientServer(Func<QuicConnection, Task> clientFunction, Func<QuicConnection, Task> serverFunction, int millisecondsTimeout = 10_000)
+        internal async Task RunClientServer(Func<QuicConnection, Task> clientFunction, Func<QuicConnection, Task> serverFunction, int iterations = 1, int millisecondsTimeout = 10_000)
         {
             using QuicListener listener = CreateQuicListener();
 
-            await new[]
+            var serverFinished = new ManualResetEventSlim();
+            var clientFinished = new ManualResetEventSlim();
+
+            for (int i = 0; i < iterations; ++i)
+            {
+                serverFinished.Reset();
+                clientFinished.Reset();
+
+                await new[]
+                {
+                    Task.Run(async () =>
+                    {
+                        using QuicConnection serverConnection = await listener.AcceptConnectionAsync();
+                        await serverFunction(serverConnection);
+                        serverFinished.Set();
+                        clientFinished.Wait();
+                        await serverConnection.CloseAsync(0);
+                    }),
+                    Task.Run(async () =>
+                    {
+                        using QuicConnection clientConnection = CreateQuicConnection(listener.ListenEndPoint);
+                        await clientConnection.ConnectAsync();
+                        await clientFunction(clientConnection);
+                        clientFinished.Set();
+                        serverFinished.Wait();
+                        await clientConnection.CloseAsync(0);
+                    })
+                }.WhenAllOrAnyFailed(millisecondsTimeout);
+            }
+        }
+
+        internal static async Task<int> ReadAll(QuicStream stream, byte[] buffer)
+        {
+            Memory<byte> memory = buffer;
+            int bytesRead = 0;
+            while (true)
+            {
+                int res = await stream.ReadAsync(memory);
+                if (res == 0)
+                {
+                    break;
+                }
+                bytesRead += res;
+                memory = memory[res..];
+            }
+
+            return bytesRead;
+        }
+
+        internal static void AssertArrayEqual(byte[] expected, byte[] actual)
+        {
+            for (int i = 0; i < expected.Length; ++i)
             {
-                Task.Run(async () =>
+                if (expected[i] == actual[i])
                 {
-                    using QuicConnection serverConnection = await listener.AcceptConnectionAsync();
-                    await serverFunction(serverConnection);
-                }),
-                Task.Run(async () =>
+                    continue;
+                }
+
+                Console.WriteLine($"Wrong data starting from pos={i}");
```

Comment:
If you need to log in test, you should rather use `ITestOutputHelper`, console output gets suppressed by xUnit.
https://github.com/dotnet/runtime/blob/6902c5accfcea6b0bee8e54cce1494c874cee266/src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTestBase.cs#L29

Reply:
`ITestOutputHelper` doesn't output to console if the tests are run from the command line... as the method I've added is more an extended assert than just some test logging, I guess it will be more useful then to use an Assert method that has a message parameter.